src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs

```diff
@@ -0,0 +1,129 @@
+using Microsoft.DotNet.Cli;
+using System.Linq;
+using System.IO;
+using Microsoft.DotNet.Tools.Common;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.DotNet.ProjectJsonMigration
+{
+    public class ProjectMigrator
+    {
+        // TODO: Migrate PackOptions
+        // TODO: Support Mappings in IncludeContext Transformations
+        // TODO: Migrate Multi-TFM projects
+        // TODO: Tests
+        // TODO: Out of Scope
+        //     - Globs that resolve to directories: /some/path/**/somedir
+        //     - Migrating Deprecated project.jsons
+        //     - Configuration dependent source exclusion
+
+        private readonly IMigrationRule _ruleSet;
+
+        public ProjectMigrator() : this(new DefaultMigrationRuleSet()) { }
+
+        public ProjectMigrator(IMigrationRule ruleSet)
+        {
+            _ruleSet = ruleSet;
+        }
+
+        public void Migrate(MigrationSettings migrationSettings)
+        {
+            var migrationRuleInputs = ComputeMigrationRuleInputs(migrationSettings);
+            VerifyInputs(migrationRuleInputs, migrationSettings);
+
+            SetupOutputDirectory(migrationSettings.ProjectDirectory, migrationSettings.OutputDirectory);
+
+            _ruleSet.Apply(migrationSettings, migrationRuleInputs);
+        }
+
+        private MigrationRuleInputs ComputeMigrationRuleInputs(MigrationSettings migrationSettings)
+        {
+            var projectContexts = ProjectContext.CreateContextForEachFramework(migrationSettings.ProjectDirectory);
+
+            var templateMSBuildProject = migrationSettings.MSBuildProjectTemplate;
+            if (templateMSBuildProject == null)
+            {
+                throw new Exception("Expected non-null MSBuildProjectTemplate in MigrationSettings");
+            }
+
+            var propertyGroup = templateMSBuildProject.AddPropertyGroup();
+            var itemGroup = templateMSBuildProject.AddItemGroup();
+
+            return new MigrationRuleInputs(projectContexts, templateMSBuildProject, itemGroup, propertyGroup);
+        }
+
+        private void VerifyInputs(MigrationRuleInputs migrationRuleInputs, MigrationSettings migrationSettings)
+        {
+            VerifyProject(migrationRuleInputs.ProjectContexts, migrationSettings.ProjectDirectory);
+        }
+
+        private void VerifyProject(IEnumerable<ProjectContext> projectContexts, string projectDirectory)
+        {
+            if (projectContexts.Count() > 1)
+            {
+                MigrationErrorCodes.MIGRATE20011($"Multi-TFM projects currently not supported.").Throw();
+            }
+
+            if (!projectContexts.Any())
+            {
+                MigrationErrorCodes.MIGRATE1013($"No projects found in {projectDirectory}").Throw();
+            }
+
+            if (projectContexts.First().LockFile == null)
+            {
+                MigrationErrorCodes.MIGRATE1012(
+                    $"project.lock.json not found in {projectDirectory}").Throw();
```

Comment:
The error message here will be `MIGRATE1012::Project Not Restored: project.lock.json not found in {projectDirectory}` to be clear. The Error code carries information on the general error category. 

I agree it should say run dotnet restore but not sure if it should be in the message or category.